src/mame/casio/cz1.cpp

```diff
@@ -0,0 +1,924 @@
+{
+	/*
+	bit 0: sound chip #1 routing (0: center, 1: left)
+	bit 1: sound chip #2 routing (0: center, 2: right)
+	bit 2: center channel stereo chorus (0: on, 1: off)
+	*/
+	m_mixer[0]->set_input_gain(1, BIT(data, 1) ? 0.0 : 1.0);
+	m_mixer[1]->set_input_gain(0, BIT(data, 0) ? 0.0 : 1.0);
+}
+
+/**************************************************************************/
+void cz1_state::cart_addr_w(u8 data)
+{
+	m_cart_addr &= 0x3f00;
+	m_cart_addr |= ~data;
+}
+
+/**************************************************************************/
+void cz1_state::cart_addr_hi_w(u8 data)
+{
+	m_cart_addr &= 0xff;
+	m_cart_addr |= (~data & 0x3f) << 8;
+}
+
+/**************************************************************************/
+void cz1_state::main_pa_w(u8 data)
+{
+	m_hd44780->db_w(data);
+	m_main_port[0] = data;
+}
+
+/**************************************************************************/
+u8 cz1_state::main_pa_r()
+{
+	u8 data = m_hd44780->db_r();
+	if (!BIT(m_main_port[1], 2))
+		data &= m_cart->read(m_cart_addr);
+	return data;
+}
+
+/**************************************************************************/
+void cz1_state::main_pb_w(u8 data)
+{
+	if (BIT(data ^ m_main_port[1], 2))
+		m_subcpu->set_input_line(UPD7810_INTF1, BIT(data, 2));
+
+	if (!BIT(data, 6) && BIT(m_main_port[1], 6) && !BIT(m_main_port[2], 2))
+		m_cart->write(m_cart_addr, m_main_port[0]);
+
+	m_hd44780->e_w(BIT(~data, 7));
+	m_hd44780->rw_w(BIT(data, 6));
+	m_hd44780->rs_w(BIT(data, 5));
+
+	m_main_port[1] = data;
+}
+
+/**************************************************************************/
+void cz1_state::main_pc_w(u8 data)
+{
+	m_main_port[2] = data;
+}
+
+
+/**************************************************************************/
+void cz1_state::sound_w(u8 data)
+{
+	m_upd933[0]->write(data);
+	m_upd933[1]->write(data);
+}
+
+/**************************************************************************/
+u8 cz1_state::sound_r()
+{
+	return m_upd933[0]->read() & m_upd933[1]->read();
+}
+
+/**************************************************************************/
+void cz1_state::sub_pa_w(u8 data)
+{
+	for (int i = 0; i < 15; i++)
+		m_led_env[i] = (BIT(data, 0, 4) == i);
+	for (int i = 0; i < 8; i++)
+		m_led_tone[i] = !BIT(data, 7) && (BIT(data, 4, 3) == i);
+}
+
+/**************************************************************************/
+void cz1_state::sub_pb_w(u8 data)
+{
+	m_upd933[0]->id_w(BIT(data, 5));
+	m_upd933[1]->id_w(BIT(data, 5));
+
+	m_upd933[0]->cs_w(BIT(data, 2));
+	m_upd933[1]->cs_w(BIT(data, 3));
+}
+
+/**************************************************************************/
+void cz1_state::sub_pc_w(u8 data)
+{
+	for (int i = 0; i < 8; i++)
+		m_led_bank[i] = !BIT(data, 0) && (BIT(data, 5, 3) == i);
+}
+
+/**************************************************************************/
+void cz1_state::main_to_sub_0_w(u8 data)
+{
+	machine().scheduler().synchronize();
+	m_main_to_sub[0] = data;
+	m_sync = 1;
+}
```

Comment:
This isn’t really going to work properly.  The call to `synchronize()` doesn’t take effect synchronously, so the new value of `m_main_to_sub[0]` will be visible before the other devices have caught up.  You need to give `synchronize()` a callback to do the state update.